nixos/modules/services/networking/cgit.nix

```diff
@@ -0,0 +1,192 @@
+{ config, lib, pkgs, ...}:
+
+with lib;
+
+let
+  cfgs = config.services.cgit;
+
+  settingType = with types; oneOf [ bool int str ];
+
+  genAttrs' = names: f: listToAttrs (map f names);
+
+  regexEscape =
+    let
+      # taken from https://github.com/python/cpython/blob/05cb728d68a278d11466f9a6c8258d914135c96c/Lib/re.py#L251-L266
+      special = [
+        "(" ")" "[" "]" "{" "}" "?" "*" "+" "-" "|" "^" "$" "\\" "." "&" "~"
+        "#" " " "\t" "\n" "\r" "\v" "\f"
+      ];
+    in
+      replaceStrings special (map (c: "\\${c}") special);
+
+  stripLocation = cfg: removeSuffix "/" cfg.nginx.location;
+
+  regexLocation = cfg: regexEscape (stripLocation cfg);
+
+  mkFastcgiPass = cfg: ''
+    ${if cfg.nginx.location == "/" then ''
+      fastcgi_param PATH_INFO $uri;
+    '' else ''
+      fastcgi_split_path_info ^(${regexLocation cfg})(/.+)$;
+      fastcgi_param PATH_INFO $fastcgi_path_info;
+    ''
+    }fastcgi_pass unix:${config.services.fcgiwrap.socketAddress};
+  '';
+
+  cgitrcLine = name: value: "${name}=${
+    if value == true then
+      "1"
+    else if value == false then
+      "0"
+    else
+      toString value
+  }";
+
+  mkCgitrc = cfg: pkgs.writeText "cgitrc" ''
+    # global settings
+    ${concatStringsSep "\n" (
+      (mapAttrsToList cgitrcLine ({ virtual-root = cfg.nginx.location; } // cfg.settings))
+      ++ (optional (cfg.scanPath != null) (cgitrcLine "scan-path" cfg.scanPath))
+      ++ [ "" "# repository settings" ]
+      ++ (concatLists
+           (mapAttrsToList
+             (url: settings:
+               [ (cgitrcLine "repo.url" url) ]
+               ++ (mapAttrsToList (name: cgitrcLine "repo.${name}") settings)
+               ++ [ "" ]
+             )
+             cfg.repos
+           )
+         )
+    )}
+    # extra config
+    ${cfg.extraConfig}
+  '';
+
+  mkCgitReposDir = cfg:
+    if cfg.scanPath != null then
+      cfg.scanPath
+    else
+      pkgs.runCommand "cgit-repos" {
+        preferLocalBuild = true;
+        allowSubstitutes = false;
+      } ''
+        mkdir -p "$out"
+        ${
+          concatStrings (
+            mapAttrsToList
+              (name: value: ''
+                ln -s ${escapeShellArg value.path} "$out"/${escapeShellArg name}
+              '')
+              cfg.repos
+          )
+        }
+      '';
+
+in
+{
+  options = {
+    services.cgit = mkOption {
+      type = types.listOf (types.submodule ({ config, ... }: {
```

Comment:
why not use types.attrsOf? with they key being the default value for nginx.virtualHost?

Reply:
This way currently gives the ability to run multiple cgit instances on the same vhost.

Reply:
you could do the same with attrsOf.
```
services.cgit."git.example.com" = { ... };
services.cgit."git.example.com2" = { nginx.virtualHost = "git.example.com"; nginx.location = "/git"; }
```

also I'm not sure if having multiple cgits on one instance is a common usecase.
attrs are generally preferred because they can be referenced again at some other point or can be overridden which is not possible with lists